### PR TITLE
[Datasets] Add logical operator for map()

### LINKED ISSUE #31937: [Datasets] Add logical operator for map()
### Description

In this issue, we should implement the logical operator for `Dataset.map()`.

### Use case

_No response_

### PR DESCRIPTION
**Pull request by c21**
Signed-off-by: Cheng Su <scnju13@gmail.com>

## Why are these changes needed?

This PR is to add logical operator for `map()`, with change:
(1).introduce `AbstractMap` abstract logical operator, that `MapBatches`, `MapRows`, etc to extend.
(2).rename existing `plan_map_batches_op` to `plan_map_op`, to make it more general.

## Related issue number

Closes https://github.com/ray-project/ray/issues/31937.
## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by ericl — APPROVED** [review]

File: python/ray/data/_internal/logical/operators/map_operator.py [hunk 1]
```diff
@@ -22,44 +23,99 @@
     from typing_extensions import Literal
 
 
-class MapBatches(LogicalOperator):
-    """Logical operator for map_batches."""
+class BaseMapLike(LogicalOperator):
```
Comment (jianoaix) [inline python/ray/data/_internal/logical/operators/map_operator.py:26]: This name looks strange, maybe LogicalMap or AbstractMap or MapInterface?

File: python/ray/data/_internal/logical/operators/map_operator.py [hunk 2]
```diff
@@ -22,44 +23,99 @@
...
+            fn_constructor_args=fn_constructor_args,
+            fn_constructor_kwargs=fn_constructor_kwargs,
+            ray_remote_args=ray_remote_args,
+        )
+        self._batch_size = batch_size
+        self._batch_format = batch_format
+        self._zero_copy_batch = zero_copy_batch
+
+
+class Map(BaseMapLike):
```
Comment (jianoaix) [inline python/ray/data/_internal/logical/operators/map_operator.py:94]: s/Map/MapRow?

**Review by jianoaix — COMMENTED** [review]
LG mostly.

File: python/ray/data/_internal/logical/operators/map_operator.py (hunk 1 above)
Comment (c21) [reply python/ray/data/_internal/logical/operators/map_operator.py]: @jianoaix - sure, changed to `AbstractMap`.

File: python/ray/data/_internal/logical/operators/map_operator.py (hunk 2 above)
Comment (c21) [reply python/ray/data/_internal/logical/operators/map_operator.py]: hmm still want to keep `Map`, to be consistent as Dataset API name, to reduce our mental overhead.

Comment (jianoaix) [reply python/ray/data/_internal/logical/operators/map_operator.py]: I don't think need to use the same as public API though.

Comment (c21) [reply python/ray/data/_internal/logical/operators/map_operator.py]: ok, renamed to `MapRows`, to be consistent with `MapBatches` operator.

Comment (jianoaix) [reply python/ray/data/_internal/logical/operators/map_operator.py]: yeah, I think the iteration API names are quite intuitive: `iter_rows` and `iter_batches`. The `map` has been leading users to take it as the primary mapping API whereas `map_batches` is recommended.

**Pull request merged**